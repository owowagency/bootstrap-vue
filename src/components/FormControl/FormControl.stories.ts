import FormControl from '.';
import {ref} from 'vue';
import {sizes} from '@/composables/useSize';
import {types} from '@/composables/useFormControl';

export default {
    title: 'Forms/Form control',
    component: FormControl,
    argTypes: {
        disabled: {
            control: {type: 'boolean'},
        },
        placeholder: {
            control: {type: 'text'},
        },
        plainText: {
            control: {type: 'boolean'},
        },
        readonly: {
            control: {type: 'boolean'},
        },
        size: {
            control: {type: 'select'},
            options: sizes,
        },
        type: {
            control: {type: 'select'},
            options: types,
        },
    },
};

const template = (args) => ({
    components: {FormControl},
    setup() {
        return {args, value: ref('')};
    },
    template: '<FormControl v-model="value" v-bind="args" />',
});

export const Default = template.bind({});

Default.args = {
    modelValue: 'I am an input',
    placeholder: 'I am a placeholder',
};

export const Autofocused = template.bind({});

Autofocused.args = {
    modelValue: 'I am autofocused',
    autofocus: true,
};

export const Disabled = template.bind({});

Disabled.args = {
    modelValue: 'I am disabled',
    disabled: true,
};

export const PlainText = template.bind({});

PlainText.args = {
    modelValue: 'Plain text',
    plainText: true,
};

export const Readonly = template.bind({});

Readonly.args = {
    modelValue: 'I am readonly',
    readonly: true,
};

export const Type = (args) => ({
    components: {FormControl},
    setup() {
        return {args, types};
    },
    template: '<FormControl v-for="t of types" class="mb-1" :placeholder="t" :type="t" v-bind="args" />',
});
