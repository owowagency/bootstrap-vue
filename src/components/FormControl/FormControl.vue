<template>
    <input
        ref="input"
        :autofocus="autofocus"
        class="form-control"
        :class="classes"
        :disabled="disabled"
        :placeholder="placeholder"
        :readonly="readonly"
        :type="type"
        :value="modelValue"
        @input="emit('update:modelValue', $event.target.value)"
    >
</template>

<script lang="ts">
import {computed, ref} from 'vue';
import {formControlProps} from '@/composables/useFormControl';
import useClasses from '@/composables/useClasses';
import useDisabled from '@/composables/useDisabled';
import useSize from '@/composables/useSize';
</script>

<script lang="ts" setup>
const props = defineProps(formControlProps);

const emit = defineEmits(['update:modelValue']);

const input = ref<HTMLInputElement>();

const {classes} = useClasses(computed(() => [
    useDisabled(props.disabled).disabledClass.value,
    props.plainText ? 'form-control-plaintext' : '',
    props.readonly ? 'readonly' : '',
    useSize(props.size, 'form-control-{0}').sizeClass.value,
]));

defineExpose({input});
</script>
