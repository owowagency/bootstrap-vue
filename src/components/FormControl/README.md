# FormControl

```vue
<FormControl name="username" v-model="username" placeholder="JohnDoe" />
```

## Imported Props

| Prop name   | Description                                  | Type                                              | Values | Default |
| ----------- | -------------------------------------------- | ------------------------------------------------- | ------ | ------- |
| autofocus   | Automatically focus on the input when loaded | boolean                                           | -      | false   |
| disabled    | Disables the component                       | boolean                                           | -      | false   |
| modelValue  | The model value                              | string                                            | -      | ''      |
| placeholder | The placeholder of the form input            | string                                            | -      | ''      |
| plainText   | Adds the `form-control-plaintext` class      | boolean                                           | -      | false   |
| readonly    | Makes the component readonly                 | boolean                                           | -      | false   |
| size        | The size of the component                    | [Size](../../composables/useSize) (string)        | -      | 'md'    |
| type        | The type of the component                    | [Type](../../composables/useFormControl) (string) | -      | 'text'  |

## Events

| Event name        | Properties                      | Description                           |
| ----------------- | ------------------------------- | ------------------------------------- |
| update:modelValue | **value** `any` - The new value | Fired when the model value is updated |
