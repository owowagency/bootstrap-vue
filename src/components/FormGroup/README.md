# FormGroup

```vue
<FormGroup v-model="username" name="username" label="Username" />
```

## Imported Props

| Prop name   | Description                                  | Type                                              | Values | Default |
| ----------- | -------------------------------------------- | ------------------------------------------------- | ------ | ------- |
| autofocus   | Automatically focus on the input when loaded | boolean                                           | -      | false   |
| disabled    | Disables the component                       | boolean                                           | -      | false   |
| id          | The id used for the component                | string                                            | -      | `uuid`  |
| modelValue  | The model value                              | string                                            | -      | ''      |
| placeholder | The placeholder of the form input            | string                                            | -      | ''      |
| plainText   | Adds the `form-control-plaintext` class      | boolean                                           | -      | false   |
| readonly    | Makes the component readonly                 | boolean                                           | -      | false   |
| size        | The size of the component                    | [Size](../../composables/useSize) (string)        | -      | 'md'    |
| type        | The type of the component                    | [Type](../../composables/useFormControl) (string) | -      | 'text'  |

## Props

| Prop name       | Description                                                                       | Type                  | Values | Default     |
| --------------- | --------------------------------------------------------------------------------- | --------------------- | ------ | ----------- |
| description     | The description of the form group<br/>_Displayed between label and input_         | string                | -      | undefined   |
| helpText        | The help text of the form group<br/>_Displayed under input_                       | string                | -      | undefined   |
| invalidFeedback | The invalid feedback of the form group<br/>_Displayed between input and helpText_ | array\|string         | -      | () =&gt; [] |
| label           | The label to display                                                              | string                | -      | undefined   |
| labelClass      | The classes that will be forwarded to the label                                   | string\|array\|object | -      | undefined   |
| validFeedback   | The valid feedback of the form group<br/>_Displayed between input and helpText_   | array\|string         | -      | () =&gt; [] |

## Slots

| Name    | Description                       | Bindings                                                                                                                                                            |
| ------- | --------------------------------- | ------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
| label   | Displays content inside the label | **label** `string` - The text of the label                                                                                                                          |
| default | Displays the input                | **id** `string` - The id of the input<br/>**is-invalid** `boolean` - Describes if the input is invalid<br/>**is-valid** `boolean` - Describes if the input is valid |

## Events

| Event name        | Properties                      | Description                           |
| ----------------- | ------------------------------- | ------------------------------------- |
| update:modelValue | **value** `any` - The new value | Fired when the model value is updated |
