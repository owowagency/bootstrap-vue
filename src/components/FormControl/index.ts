import FormControl from './FormControl.vue';

export default FormControl;
