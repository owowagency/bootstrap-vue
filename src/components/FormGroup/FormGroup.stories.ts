import FormControl from '@/components/FormControl';
import FormGroup from '.';
import {ref} from 'vue';

export default {
    title: 'Forms/Form group',
    component: FormGroup,
    argTypes: {
        ...FormControl.argTypes,
        helpText: {
            control: {type: 'text'},
        },
        invalidFeedback: {
            control: {type: 'text'},
        },
        label: {
            control: {type: 'text'},
        },
        validFeedback: {
            control: {type: 'text'},
        },
    },
};

const template = (args) => ({
    components: {FormGroup},
    setup() {
        args.label = args.label || 'I am a form group';

        return {args, value: ref('')};
    },
    template: '<FormGroup v-model="value" v-bind="args" />',
});

export const Default = template.bind({});

export const Description = template.bind({});

Description.args = {
    description: 'This is some description',
};

export const HelpText = template.bind({});

HelpText.args = {
    helpText: 'This is some help text',
};

export const InvalidFeedback = template.bind({});

InvalidFeedback.args = {
    invalidFeedback: 'This is some invalid feedback',
};

export const ValidFeedback = template.bind({});

ValidFeedback.args = {
    validFeedback: 'This is some valid feedback',
};
