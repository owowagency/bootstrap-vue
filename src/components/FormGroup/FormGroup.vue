<template>
    <div>
        <label
            v-if="!!$slots.label || label || description"
            class="form-label"
            :class="labelClass"
            :for="id"
        >
            <!--
            @slot Displays content inside the label
            @binding {string} label The text of the label
            -->
            <slot
                name="label"
                :label="label"
            >
                {{ label }}
            </slot>

            <div
                v-if="!!description"
                class="form-text"
            >
                {{ description }}
            </div>
        </label>

        <!--
        @slot Displays the input
        @binding {string} id The id of the input
        @binding {boolean} is-invalid Describes if the input is invalid
        @binding {boolean} is-valid Describes if the input is valid
        -->
        <slot
            :id="id"
            :is-invalid="isInvalid"
            :is-valid="isValid"
        >
            <FormControl
                v-bind="propsFormControl"
                :id="id"
                :class="validityClass"
                @update:model-value="emit('update:modelValue', $event)"
            />
        </slot>

        <template v-if="isInvalid">
            <div
                v-for="feedback of invalidFeedback"
                :key="feedback"
                class="invalid-feedback"
            >
                {{ feedback }}
            </div>
        </template>

        <template v-if="isValid">
            <div
                v-for="feedback of validFeedback"
                :key="feedback"
                class="valid-feedback"
            >
                {{ feedback }}
            </div>
        </template>

        <div
            v-if="helpText"
            class="form-text"
        >
            {{ helpText }}
        </div>
    </div>
</template>

<script lang="ts">
import FormControl from '@/components/FormControl';
import {computed} from 'vue';
import extractKeysFrom from '@/library/extractKeysFrom';
import {formControlProps} from '@/composables/useFormControl';
import {idProps} from '@/composables/useId';
</script>

<script lang="ts" setup>
const props = defineProps({
    /**
     * The description of the form group
     * *Displayed between label and input*
     */
    description: {
        type: String,
        default: undefined,
    },
    /**
     * The help text of the form group
     * *Displayed under input*
     */
    helpText: {
        type: String,
        default: undefined,
    },
    ...idProps,
    ...formControlProps,
    /**
     * The invalid feedback of the form group
     * *Displayed between input and helpText*
     */
    invalidFeedback: {
        type: [Array, String],
        default: () => [],
    },
    /**
     * The label to display
     */
    label: {
        type: String,
        default: undefined,
    },
    /**
     * The classes that will be forwarded to the label
     */
    labelClass: {
        type: [String, Array, Object],
        default: undefined,
    },
    /**
     * The valid feedback of the form group
     * *Displayed between input and helpText*
     */
    validFeedback: {
        type: [Array, String],
        default: () => [],
    },
});

const emit = defineEmits([
    /**
     * Fired when the model value is updated
     * @param {any} value The new value
     */
    'update:modelValue',
]);

const propsFormControl = extractKeysFrom(Object.keys(formControlProps), props);

const invalidFeedback = computed(() => [props.invalidFeedback].flat().filter(f => !!f));

const isInvalid = computed(() => !!invalidFeedback.value.length);

const validFeedback = computed(() => [props.validFeedback].flat().filter(f => !!f));

const isValid = computed(() => !!validFeedback.value.length);

const validityClass = computed(() => [
    isInvalid.value ? 'is-invalid' : undefined,
    isValid.value ? 'is-valid' : undefined,
]);
</script>

<docs>
```vue
<FormGroup
    v-model="username"
    name="username"
    label="Username"
/>
```

## Imported Props

| Prop name   | Description                                  | Type                                              | Values | Default |
| ----------- | -------------------------------------------- | ------------------------------------------------- | ------ | ------- |
| autofocus   | Automatically focus on the input when loaded | boolean                                           | -      | false   |
| disabled    | Disables the component                       | boolean                                           | -      | false   |
| id          | The id used for the component                | string                                            | -      | `uuid`  |
| modelValue  | The model value                              | string                                            | -      | ''      |
| placeholder | The placeholder of the form input            | string                                            | -      | ''      |
| plainText   | Adds the `form-control-plaintext` class      | boolean                                           | -      | false   |
| readonly    | Makes the component readonly                 | boolean                                           | -      | false   |
| size        | The size of the component                    | [Size](../../composables/useSize) (string)        | -      | 'md'    |
| type        | The type of the component                    | [Type](../../composables/useFormControl) (string) | -      | 'text'  |
</docs>
