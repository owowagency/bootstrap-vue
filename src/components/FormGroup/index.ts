import FormGroup from './FormGroup.vue';

export default FormGroup;
