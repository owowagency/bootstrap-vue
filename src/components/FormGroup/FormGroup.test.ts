import {describe, expect, it} from 'vitest';
import FormControl from '@/components/FormControl';
import FormGroup from '.';
import {shallowMount} from '@vue/test-utils';

describe('template', () => {
    componentRenderTest(FormGroup, {
        // Force an id to prevent snapshots from failing because of uuid.
        props: {id: 'dont-change-please'},
    });

    componentSlotRenderTest(FormGroup);

    componentSlotRenderTest(FormGroup, 'label');

    it('renders scoped slot', async() => {
        const wrapper = shallowMount(FormGroup, {
            props: {
                invalidFeedback: 'a',
                validFeedback: 'a',
            },
            slots: {
                default: `
                    <template #default="params">
                        {{ params.isInvalid ? 'I am invalid' : '' }}
                        {{ params.isValid ? 'I am valid' : '' }}
                    </template>
                `,
            },
        });

        expect(wrapper.text()).toContain('I am invalid');

        expect(wrapper.text()).toContain('I am valid');
    });

    it('renders label when given', () => {
        const wrapper = shallowMount(FormGroup, {
            props: {
                id: 'some-id',
                label: 'I am a label',
            },
        });

        const label = wrapper.find('.form-label');

        expect(label.text()).toBe('I am a label');

        expect(label.attributes('for')).toBe('some-id');
    });

    it('renders labelClass when given', () => {
        const wrapper = shallowMount(FormGroup, {
            props: {
                // Add label so it will be rendered.
                label: 'I am a label',
                labelClass: 'text-primary',
            },
        });

        const label = wrapper.find('.form-label');

        expect(label.classes()).toContain('text-primary');
    });

    it('renders description when given', () => {
        const wrapper = shallowMount(FormGroup, {
            props: {description: 'I am a description'},
        });

        const description = wrapper.find('.form-label .form-text');

        expect(description.text()).toBe('I am a description');
    });

    it('adds validatity class to form control', () => {
        const wrapper = shallowMount(FormGroup, {
            props: {
                invalidFeedback: 'erroooooreeee',
                validFeedback: 'well done!',
            },
        });

        const formControl = wrapper.findComponent(FormControl);

        expect(formControl.classes()).toStrictEqual(['is-invalid', 'is-valid']);
    });

    it('emits update model value when form control emits it', () => {
        const wrapper = shallowMount(FormGroup);

        const formControl = wrapper.findComponent(FormControl);

        formControl.vm.$emit('update:modelValue', 'hello');

        expect(wrapper.emitted('update:modelValue')[0]).toStrictEqual(['hello']);
    });

    it('renders invalid feedback', () => {
        const wrapper = shallowMount(FormGroup, {
            props: {invalidFeedback: 'errooore'},
        });

        const invalidFeedback = wrapper.find('.invalid-feedback');

        expect(invalidFeedback.text()).toBe('errooore');
    });

    it('renders an array of invalid feedback', () => {
        const wrapper = shallowMount(FormGroup, {
            props: {invalidFeedback: ['errooore 1', 'errooore 2']},
        });

        expect(wrapper.findAll('.invalid-feedback').length).toBe(2);
    });

    it('renders valid feedback', () => {
        const wrapper = shallowMount(FormGroup, {
            props: {validFeedback: 'well done!'},
        });

        const validFeedback = wrapper.find('.valid-feedback');

        expect(validFeedback.text()).toBe('well done!');
    });

    it('renders an array of valid feedback', () => {
        const wrapper = shallowMount(FormGroup, {
            props: {validFeedback: ['well', 'done']},
        });

        expect(wrapper.findAll('.valid-feedback').length).toBe(2);
    });

    it('renders help text', () => {
        const wrapper = shallowMount(FormGroup, {
            props: {helpText: 'some more info'},
        });

        const helpText = wrapper.find('.form-text');

        expect(helpText.text()).toBe('some more info');
    });
});

describe('invalidFeedback', () => {
    it('always is an array', () => {
        const wrapper = shallowMount(FormGroup, {
            props: {invalidFeedback: 'string'},
        });

        expect(wrapper.vm.invalidFeedback).toEqual(['string']);
    });

    it('filters', () => {
        const wrapper = shallowMount(FormGroup, {
            props: {invalidFeedback: ['', 0, undefined]},
        });

        expect(wrapper.vm.invalidFeedback).toEqual([]);
    });
});

describe('isInvalid', () => {
    it('is false when has no invalid feedback', () => {
        const wrapper = shallowMount(FormGroup, {
            props: {invalidFeedback: ''},
        });

        expect(wrapper.vm.isInvalid).toBe(false);
    });

    it('is true when has invalid feedback', () => {
        const wrapper = shallowMount(FormGroup, {
            props: {invalidFeedback: 'erroooreeee'},
        });

        expect(wrapper.vm.isInvalid).toBe(true);
    });
});


describe('validFeedback', () => {
    it('always is an array', () => {
        const wrapper = shallowMount(FormGroup, {
            props: {validFeedback: 'string'},
        });

        expect(wrapper.vm.validFeedback).toEqual(['string']);
    });

    it('filters', () => {
        const wrapper = shallowMount(FormGroup, {
            props: {validFeedback: ['', 0, undefined]},
        });

        expect(wrapper.vm.validFeedback).toEqual([]);
    });
});

describe('isValid', () => {
    it('is false when has no valid feedback', () => {
        const wrapper = shallowMount(FormGroup, {
            props: {validFeedback: ''},
        });

        expect(wrapper.vm.isValid).toBe(false);
    });

    it('is true when has valid feedback', () => {
        const wrapper = shallowMount(FormGroup, {
            props: {validFeedback: 'well done'},
        });

        expect(wrapper.vm.isValid).toBe(true);
    });
});
