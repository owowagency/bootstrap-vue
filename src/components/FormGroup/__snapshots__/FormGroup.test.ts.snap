// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`template renders default 1`] = `
"<div>
  <!--v-if-->
  <form-control-stub autofocus=\\"false\\" disabled=\\"false\\" modelvalue=\\"\\" placeholder=\\"\\" plaintext=\\"false\\" size=\\"md\\" type=\\"text\\" id=\\"dont-change-please\\" class=\\"\\"></form-control-stub>
  <!--v-if-->
  <!--v-if-->
  <!--v-if-->
</div>"
`;
